src/gamma.jl

```diff
@@ -615,7 +614,31 @@ the upper incomplete gamma function ``\Gamma(a,z)``.
 - `Complex`: by `exp(loggamma(z))`.
 - `BigFloat`: C library for multiple-precision floating-point [MPFR](https://www.mpfr.org/)
 """
-gamma(x::Number) = gamma(float(x))
+function gamma(x::Number)
+    z = float(x)
+    if typeof(z) === typeof(x)
+        # There is nothing to fallback to
+        throw(MethodError(gamma, x))
+    end
+    return gamma(z)
+end
```

Comment:
The way I'd prefer to fix this is to do:
```jl
gamma(x::Number) = _gamma(float(x))
```
and then define methods for `_gamma`.  That way we will get a `MethodError` automatically for unsupported types.

Reply:
Ah OK, I assumed the approach in this PR would be preferred since it is already used for `digamma` and other functions: https://github.com/JuliaMath/SpecialFunctions.jl/blob/74fd16fb694b893bb008f352d9a9c17342293c99/src/gamma.jl#L526-L531

But if this is not the case I'll update everything, including `digamma` etc. to the simpler dispatch-based approach.

Reply:
The function implementations were developed at various times and were reviewed different groups of people, e.g. some of them came from Base, which is why there is probably more diversity in coding styles than in a typical package.  It would great to make this more uniform.